src/main/kotlin/com/github/izhangzhihao/rainbow/brackets/indents/RainbowIndentsPass.kt

```diff
@@ -0,0 +1,507 @@
+        for (line in 1 until lineIndents.size) {
+            ProgressManager.checkCanceled()
+            val curIndent = abs(lineIndents[line])
+
+            while (!indents.empty() && curIndent <= indents.peek()) {
+                ProgressManager.checkCanceled()
+                val level = indents.pop()
+                val startLine = lines.pop()
+                if (level > 0) {
+                    for (i in startLine until line) {
+                        if (level != abs(lineIndents[i])) {
+                            descriptors.add(createDescriptor(level, startLine, line, lineIndents))
+                            break
+                        }
+                    }
+                }
+            }
+
+            val prevLine = line - 1
+            val prevIndent = abs(lineIndents[prevLine])
+
+            if (curIndent - prevIndent > 1) {
+                lines.push(prevLine)
+                indents.push(prevIndent)
+            }
+        }
+
+        val document = checkNotNull(myDocument)
+        while (!indents.empty()) {
+            ProgressManager.checkCanceled()
+            val level = indents.pop()
+            val startLine = lines.pop()
+            if (level > 0) {
+                descriptors.add(createDescriptor(level, startLine, document.lineCount, lineIndents))
+            }
+        }
+        return descriptors
+    }
+
+    private fun createDescriptor(
+            level: Int,
+            startLine: Int,
+            endLine: Int,
+            lineIndents: IntArray
+    ): IndentGuideDescriptor {
+        var sLine = startLine
+        while (sLine > 0 && lineIndents[sLine] < 0) sLine--
+        // int codeConstructStartLine = findCodeConstructStartLine(startLine);
+        return IndentGuideDescriptor(level, sLine, endLine)
+    }
+
+    /*
+    private fun findCodeConstructStart(startLine: Int): Int? {
+        val document = myEditor.document
+        val text = document.immutableCharSequence
+        val lineStartOffset = document.getLineStartOffset(startLine)
+        val firstNonWsOffset = CharArrayUtil.shiftForward(text, lineStartOffset, " \t")
+        val type = PsiUtilBase.getPsiFileAtOffset(myFile, firstNonWsOffset).fileType
+        val language = PsiUtilCore.getLanguageAtOffset(myFile, firstNonWsOffset)
+        val braceMatcher = BraceMatchingUtil.getBraceMatcher(type, language)
+        val iterator = myEditor.highlighter.createIterator(firstNonWsOffset)
+        return if (braceMatcher.isLBraceToken(iterator, text, type)) {
+            braceMatcher.getCodeConstructStart(myFile, firstNonWsOffset)
+        } else null
+    }
+
+
+    private fun findCodeConstructStartLine(startLine: Int): Int {
+        val codeConstructStart = findCodeConstructStart(startLine)
+        return if (codeConstructStart != null) myEditor.document.getLineNumber(codeConstructStart) else startLine
+    }
+    */
+
+    private inner class IndentsCalculator internal constructor() {
```

Comment:
We won't provide public API as IntelliJ, so `IndentsCalculator` could be public and move to another file I think.

Reply:
After that, once Intellij make `IndentsCalculator` public, our `IndentsCalculator` can become retirement.

Reply:
`IndentsCalculator` is unlikely to be made public.